lib/puppet/application/plugin.rb

```diff
@@ -1,3 +1,4 @@
 require 'puppet/application/face_base'
 class Puppet::Application::Plugin < Puppet::Application::FaceBase
+  environment_mode :not_required
```

Comment:
Don't we need to know which environment to download the plugins to? Where do these land if the environment isn't required?

Reply:
We request plugins from an environment on the master, but always download into the same `Puppet[:libdir]` directory. So it's not important if the environment we're requesting plugins from exists locally or not.

It's only in cases where we are running in a compiler-context (like puppetserver, puppet lookup, puppet apply) that we absolutely need the environment to exist. It would have been nice if `Puppet[:environment]` wasn't overloaded for both compiler code and non-compiler code, but that 🚢 has sailed.

Reply:
Ah I left out some important words above:

>  So it's not important if the environment we're requesting plugins from exists locally or not.

It's not important **to the agent** if the environment we're requesting plugins from exists locally **on the agent** or not. It is important that the environment exists on the master so that there's a directory from which the master can serve file metadata and content. That said, this test doesn't cover master behavior, just agent behavior. There are other beaker tests that verify environment handling when puppet ruby code runs in puppetserver.

Reply:
Ah, okay, that's what I was wondering about. So there are checks in place elsewhere that ensure the environment exists on the master?

Reply:
Yep, yep. The server side REST handler calls into https://github.com/puppetlabs/puppet/blob/master/lib/puppet/network/http/api/indirected_routes.rb#L96-L111 which pulls the environment off of the HTTP query parameter or defaults to the environment the server is configured with. Then later it calls into the indirector, creating a `Puppet::Indirector::Request` and that verifies the environment exists: https://github.com/puppetlabs/puppet/blob/b16ea6ee1f7f154f994b641eb380087584928534/lib/puppet/indirector/request.rb#L37-L44. It's a little convoluted, but the `set_attributes` method automatically calls the setter methods based on https://github.com/puppetlabs/puppet/blob/b16ea6ee1f7f154f994b641eb380087584928534/lib/puppet/indirector/request.rb#L23

Reply:
Cool!